extensions/razer/internal.m

```diff
@@ -0,0 +1,1658 @@
+            NSString *filename = (NSString *)obj;
+            NSString *extension = [[filename pathExtension] lowercaseString];
+            if ([extension isEqualToString:@"json"]) {
+                [jsonFiles addObject:[jsonFilePath stringByAppendingPathComponent:filename]];
+            }
+        }];
+        
+        NSMutableDictionary *devices = [NSMutableDictionary dictionary];
+        
+        for (NSString *filePath in jsonFiles) {
+            NSData *JSONData = [NSData dataWithContentsOfFile:filePath options:NSDataReadingMappedIfSafe error:NULL];
+            
+            NSDictionary *JSONObject = [NSJSONSerialization
+                             JSONObjectWithData:JSONData
+                             options:NSJSONReadingAllowFragments
+                             error:NULL];
+            
+            NSString *productId = [JSONObject objectForKey:@"productId"];
+                        
+            [devices setObject:JSONObject forKey:productId];
+        }
+        devicesCache = devices;
+    }
+    return devicesCache;
+}
+
+#pragma mark - Support Functions and Classes
+
+@interface HSRazer : NSObject
+
+@property (nonatomic, strong) id    ioHIDManager;
+
+@property CFMachPortRef             eventTap;
+
+@property NSNumber*                 internalDeviceId;
+@property NSNumber*                 productId;
+
+@property NSString*                 name;
+@property NSString*                 mainType;
+@property NSString*                 image;
+@property NSMutableArray*           features;
+@property NSMutableArray*           featuresConfig;
+@property NSMutableArray*           featuresMissing;
+
+@property BOOL                      scrollWheelPressed;
+@property BOOL                      scrollWheelInProgress;
+
+@property int                       selfRefCount;
+@property int                       callbackRef;
+@property id                        callbackToken;
+@property int                       deviceCallbackRef;
+
+@property LSGCCanary                lsCanary;
+
+@end
+
+@implementation HSRazer
+
+- (instancetype)init
+{
+    self = [super init];
+    if (self)
+    {
+        _callbackRef                = LUA_NOREF;
+        _deviceCallbackRef          = LUA_NOREF;
+        _callbackToken              = nil;
+        _selfRefCount               = 0;
+        _scrollWheelPressed         = NO;
+        _scrollWheelInProgress      = NO;
+    }
+    return self;
+}
+
+#pragma mark - Event Tap for Scroll Wheel
+
+// We need to use an eventtap to stop the scroll wheel on the Razer Tartarus V2 from actually scrolling.
+
+static CGEventRef eventTapCallback(CGEventTapProxy proxy,
+                           CGEventType type,
+                           CGEventRef event,
+                           void* refcon) {
+    
+    // Prevent a crash when doing garbage collection:
+    if (type == kCGEventTapDisabledByUserInput) {
+        return event;
+    }
+        
+    HSRazer *manager = (__bridge HSRazer *)refcon;
```

Comment:
I think you should check if `manager` is `nil`. It's probably not something that would cause a crash further down, but best to be safe I think.

Reply:
Actually, I'd love your advice on this anyway. I feel like there's something I'm very much misunderstanding anyway. I did try checking if `manager` is `nil`, but it still crashed during garbage collection because manager was a `NSZombie`. I feel like I'm missing something about using a bridge?